_blog/2022-12-23-procedural-macros.mdx

```diff
@@ -0,0 +1,461 @@
+
+### Procedural macro time!
+
+Now we know what code we want to generate, we can write some Rust code to handle a structure and generate the an output token stream.
+
+Rust procedural macros require their own crate for their definition due to constraints on how they are compiled. The unofficial Rust convention for derive macros is the name of trait or crate name + derive. So let's run the following from our current folder `cargo new --lib binary-serialize-derive`. We need to let `cargo` know that this crate is a proc macro by defining it in its `Cargo.toml`. We'll also add the dependencies [syn](https://crates.io/crates/syn) for parsing the contents of our structure and [quote](https://crates.io/crates/quote) for generating the output:
+
+```toml
+[package]
+name = "binary-serialize-derive"
+version = "0.1.0"
+edition = "2021"
+
+[lib]
+# Important \/\/\/
+proc-macro = true
+
+[dependencies]
+# Dependencies we use when writing the macro \/\/\/
+quote = "1.0.23"
+syn = "1.0.107"
+```
+
+In our macro we want to parse the input into a structure that we can read information from. In the below we can read `.fields` directly without having to understand what tokens refer to field names and such.
+```rust
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::{parse_macro_input, DeriveInput, Data};
+
+#[proc_macro_derive(BinarySerializable)]
+pub fn my_macro(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+
+    // Get the name of the structure the derive is on. For example if we have:
+    // ```rust
+    // #[derive(BinarySerializable)]
+    // struct MyStruct;
+    // ```
+    // `name` will be "MyStruct"
+    let name = input.ident;
+
+    // Only look at struct data for now.
+    // Challenge for the reader to continue the logic for enums!
+    let Data::Struct(struct_data) = input.data else {
+        unimplemented!("enums");
+    };
+
+    // Produce a iterator of expressions for serializing each field
+    let serialize_fields = struct_data.fields.iter().map(|field| {
+        let Some(ident) = &field.ident else {
+            unimplemented!("tuple structs");
+        };
+        quote!( self.#ident.serialize(buf); )
+    });
+
+    // Produce a key value pair deserializing each field
+    // VERY important that the iterator is in the same order as the above serialize iterator!
+    let deserialize_fields = struct_data.fields.iter().map(|field| {
+        let Some(ident) = &field.ident else {
+            unimplemented!("tuple structs");
+        };
+        quote!( #ident: BinarySerializable::deserialize(iter) )
+    });
+
+    // Finally produce a `impl` block for our trait. Using `#` to interpolate are above token iterators
+    let expanded = quote! {
+        #[automatically_derived]
+        impl BinarySerializable for #name {
+            fn serialize(self, buf: &mut Vec<u8>) {
+                #( #serialize_fields );*
```

Comment:
```suggestion
                #( #serialize_fields )*
```

`serialize_fields` already added the `;`. Will have to console `deserialize_fields` to be the same